### ISSUE TITLE
How to reproduce painn results on qm9 and md17?

### ISSUE THREAD

**Issue by Nourollah** [atomistic-machine-learning/schnetpack#672]
Hi,

I am having trouble reproducing the QM9 and MD17 results in the painn paper (Schutt et. al. 2021). Is there a standard way I can do this? I have managed to reproduce the QM9 results original repository including with the paper on the ICML website.

The code in the tutorials (Jupyter notebooks) doesn't even seem to approximately reproduce the values for both datasets.

I would really appreciate some help.

Best wishes,
Masoud

**Comment by jnsLs** [comment]
Hi Masoud,

I would recommend using the CLI to reproduce the benchmarks. 
To train a model on QM9, you can simply use the command 'spktrain experiment=qm9_atomwise'.
Similarly you can train models on MD17. It is possible that some default hyperparameters might have changed by now. So please make sure that the hyperparameters in the 'config.yaml' file are identical to those used in the paper.

Best regards, 
Jonas

**Comment by Nourollah** [comment]
Hello Jonas,

I spotted the thread [this thread](https://github.com/atomistic-machine-learning/schnetpack/issues/77) and am aware that I can utilize the CLI to replicate the benchmarks. However, I am encountering difficulties in reproducing the benchmarks directly through the code. I have verified the hyperparameters, and they align with those presented in the paper. If there exists a discrepancy between the code and the CLI, I would appreciate insights into the potential causes of this issue. Additionally, I am curious about the possibility of incorporating another model into the project for comparative analysis with PainNN or SchNet.

In my opinion, there should be a mechanism to replicate the CLI-generated results through the code or Jupyter Notebooks.

Warm regards,
Masoud

**Comment by Nourollah** [comment]
Hello,

I am updating this thread. I have double-checked everything, and it appears that I am reproducing the results for MD17 correctly. However, I obtained better results than the ones reported in the paper, which used three different random seeds. This suggests that I may be making a mistake somewhere or that the project has been updated, and the hyperparameters or models are now working more effectively due to a change in the version of PyTorch or another factor. For QM9, it appears that the data is slightly different from the data that I can use with PyTorch-Geometric or the Schnet repository, which reported on the paper.

**Comment by stefaanhessmann** [comment]
Hi @Nourollah,

I am happy to hear that you could reproduce the results of the paper. Of course you can do this without the CLI, but one needs to be careful to set all hyperparameters correctly.
Differences in performance (usually providing better results) can still occur, because some default parameters have changed in SchNetPack. This mainly affects the activation functions, number of interaction blocks, feature dimensions and also  other training parameters.

What exactly do you mean by differences in the QM9 dataset? The dataset has an argument `remove_uncharacterized` which lets you filter out invalid structures. This might cause the differences.

Best,
Stefaan

**Comment by Nourollah** [comment]
Hello @stefaanhessmann,

Thank you for your detailed explanation. I concur that both methods utilize the same base code, leading to the expectation of similar results. However, upon reviewing the hyperparameters you mentioned, they are remarkably similar to those presented in the paper. Furthermore, I am employing the same code for QM9 and MD17. Consequently, I anticipate obtaining close results to the paper for QM9, similar to those obtained for MD17 (which is marginally superior to the paper and I believe this discrepancy should not be attributed to the identical hyperparameters used).

Following your suggestion about `remove_uncharacterized` portion of the dataset, I did experiments to train the PaiNN with and without this subset. Based on the obtained results, I discern no significant difference between the two approaches. Notably, the results obtained are substantially inferior to those presented in the paper (approximately 10 times worse with seed 2000 and 12 times worse with seed 42). I have ensured that the number of interaction blocks, activation functions, and feature dimensions aligns precisely with the repo addressed in the PaiNN paper (Schutt et al., 2021).

Upon plotting a chart depicting the mean and standard deviation of U0 for the QM9 dataset in conjunction with the SchNetPack dataset and the PyTorch geometric dataset, I observed distinct differences. I speculate that this discrepancy may be the root cause of the issue. Additionally, I attempted to modify the units, but this returned no noticeable change to reproducing the paper results.

**Comment by stefaanhessmann** [comment]
Ok, we will further look into this. If the model is worse by a factor for 10 or 12, it might be related to some issues with unit conversions? How do you evaluate the test data? Did you check, if maybe some unit conversion from kcal/mol to eV or vice versa.
Could you maybe add the exact test results that you observed? And what version of SchNetPack did you use?

**Comment by jnsLs** [comment]
Hi @Nourollah,
we are able to roughly reproduce the results from the PaiNN-paper. 
Any updates from your side? Maybe looking into the closed issue #674 might help?
Best regards, 
Jonas

**Comment by Nourollah** [comment]
Hi @jnsLs ,

Unfortunately, there has been no progress. I was unable to reproduce the results from the paper, nor have I been able to obtain results that are even close when optimizing all the hyperparameters.

Thank you for your comment, @stefaanhessmann . I ran several experiments with different units to investigate whether that might be the issue. The units appear to be correct. For example, I obtained a value of `114.28 meV` for U0 in my test with seed 42.

**Comment by Nourollah** [comment]
#Hi @jnsLs ,

Thank you for your support. I understand that the tutorial is not intended to provide a fully configured model with configuration files. However, I would like to ensure that I have a correct understanding of the accuracy demonstrated in the tutorial, as this will help me effectively use the project for my purposes.
The code I am using is almost this:

```python
import os
import schnetpack as spk
from schnetpack.datasets import QM9
import schnetpack.transform as trn

import torch
import torchmetrics
import pytorch_lightning as pl

import wandb

qm9tut = './qm9tut'
if not os.path.exists('qm9tut'):
    os.makedirs(qm9tut)

qm9data = QM9(
    './qm9.db', 
    batch_size=100,
    num_train=109000,
    num_val=10000,
    num_test=10000,
    transforms=[
        trn.ASENeighborList(cutoff=5.),
        trn.RemoveOffsets(QM9.U0, remove_mean=True, remove_atomrefs=False),
        trn.CastTo32()
    ],
    property_units={QM9.U0: 'eV'},
    num_workers=4,
    split_file=os.path.join(qm9tut, "split.npz"),
    pin_memory=True, 
    load_properties=[QM9.U0],
)
qm9data.prepare_data()
qm9data.setup()

atomrefs = qm9data.train_dataset.atomrefs
print('U0 of hyrogen:', atomrefs[QM9.U0][1].item(), 'eV')
print('U0 of carbon:', atomrefs[QM9.U0][6].item(), 'eV')
print('U0 of oxygen:', atomrefs[QM9.U0][8].item(), 'eV')

means, stddevs = qm9data.get_stats(
    QM9.U0, divide_by_atoms=True, remove_atomref=False
)
print('Mean atomization energy / atom:', means.item())
print('Std. dev. atomization energy / atom:', stddevs.item())

cutoff = 5.
n_atom_basis = 128

pairwise_distance = spk.atomistic.PairwiseDistances() 
radial_basis = spk.nn.GaussianRBF(n_rbf=20, cutoff=cutoff)
schnet = spk.representation.SchNet(
    n_atom_basis=n_atom_basis, n_interactions=3,
    radial_basis=radial_basis,
    cutoff_fn=spk.nn.CosineCutoff(cutoff)
)
pred_U0 = spk.atomistic.Atomwise(n_in=n_atom_basis, output_key=QM9.U0)

nnpot = spk.model.NeuralNetworkPotential(
    representation=schnet,
    input_modules=[pairwise_distance],
    output_modules=[pred_U0],
    postprocessors=[trn.CastTo64(), trn.AddOffsets(QM9.U0, add_mean=True, add_atomrefs=False)]
)

output_U0 = spk.task.ModelOutput(
    name=QM9.U0,
    loss_fn=torch.nn.MSELoss(),
    loss_weight=1.,
    metrics={
        "MAE": torchmetrics.MeanAbsoluteError(),
	    "RMSE": torchmetrics.regression.MeanSquaredError(squared=False)
    }
)

task = spk.task.AtomisticTask(
    model=nnpot,
    outputs=[output_U0],
    optimizer_cls=torch.optim.AdamW,
    optimizer_args={"lr": 5e-4}
)

logger = pl.loggers.TensorBoardLogger(save_dir=qm9tut)
wandb.init(
    project="PaiNN",
    name="Reproducing_paper_result",
    sync_tensorboard=True
)

callbacks = [
    spk.train.ModelCheckpoint(
        model_path=os.path.join(qm9tut, "best_inference_model"),
        save_top_k=1,
        monitor="val_loss"
    ),
	pl.callbacks.EarlyStopping(
		monitor="val_loss",
		min_delta=0.005,
		patience=30,
		verbose=False,
		mode="min"
    )
]

trainer = pl.Trainer(
    callbacks=callbacks,
    logger=logger,
    default_root_dir=qm9tut,
    max_epochs=10000,
)
trainer.fit(task, datamodule=qm9data)
trainer.test(task, qm9data.test_dataloader())
```

>

**Comment by jnsLs** [comment]
Hi @Nourollah,

`remove_atomrefs` must be set to `True`. In your case its `False`.
Be sure that the file at this location is removed before setting up the dataset. A common problem is that old split files at the `split_file` path are loaded.

**Comment by Nourollah** [comment]
Hi @jnsLs ,

Thank you for reviewing the code. I have attempted this before and am aware of the database file’s structure. While it is created only once, rebuilding the database with new changes is necessary. However, this is not the issue at hand, and I am obtaining various results with different parameter settings and neither are close to the paper results.

**Comment by jnsLs** [comment]
- So you have tried `remove_atomrefs=True`? Setting it to `False` will cause numerical problems.
- Also I would recommend using all data by setting `num_train=110000` and `num_val=10000` and `num_test=None`. In case you are not using the latest SchNetPack commit `num_test=None` is very important.
- If you want to reproduce the results from the painn paper you should use the `representation=painn`.
- Looks like you are not using a learning rate scheduler. This is problematic.
- you are neither using `exponential mean average`, which is less problematic but they used it in the paper.

**Comment by Nourollah** [comment]
Despite implementing all of your suggestions, the results still deviate significantly from the paper. I was able to reproduce the value of `29.9meV` for U0. I attempted to adhere to all the configurations for PAINN specified in the configuration files.
